Add NodePool support to Cluster resource

birdayz opened the pull request:
# Operator v1: support for multiple Node Pools



Overall design notes
- Everything is driven by spec.nodePools. The default nodepool, driven by the existing fields spec.replicas, spec.storage etc, is still fully supported. It is automatically injected into internal data structures as a nodePool with the name "default". This way. the code needs (almost) no special treatment for the default nodepool, but still supports it in full.
- Deleted nodePools - so nodepools removed from spec.nodePools are handled by detecting their statefulset, and injecting a synthetic nodepoolSpec into the internal data structures.
- Replace AdminAPIFactory with a factory that works with pods, instead of ordinals - as we need to support pods with different prefixes (nodepools).
- Reduce cross-nodepool dependencies as much as possible, try to avoid code that requires overall replica count if possible. It is not possible everywhere, as we support for example only one pod in decomission at the same time.
- Cloud Storage percentage has been reworked. We do not want to set bytes-specific cluster prop anymore. Instead, only use percentage based property.
  This avoids problems with multiple node pools: bytes-based is very different if disk size changes (move between different tiers in cloud, for example).Percentage based is kept stable mostly (even exactly the same most of the time, 15% typically), and will not cause immediate churn on the old nodePool if disk size is going down.
- Previously, if a brokerID is marked for decomission (status.decomissionBrokerID), but eventually the pod for the brokerID is not found, the decommission was flagged as done, and the status field cleared. However, this is not safe: if this happens, in a race condition, the broker may still get registered - but is not being decommissioned anymore, making it a ghost broker.
- Made featuregate EmptySeedStartCluster mandatory.
  Clusters <22.3.0 are not supported anymore. This significantly reduces complexity
  with multi-nodepool-support. We would have to start the first node in initial startup, 
  and wait for this node. However, with multiple nodepools, all nodepools would have to
  coordinate this. Since this is a very old version, that is not supported anymore, 
  support if therefore removed to simplify.

Not directly related changes, to reduce test flakiness:
- Changed tests to require less resources, down from 1G,1cpu. Some tests run longer - if two tests like this run concurrently (we have concurrency limit 2), one of them sometimes gets starved on cpu/mem and fails to schedule until timeout.
- Bumped timeout of readiness probe to 5s (had none configured => default 1s). this helped making lower resource allocations work. It is also rarely a problem in production, so IMHO no-brainer to bump it a little.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redpanda-data/redpanda-operator?pullRequest=188) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/redpanda-data/redpanda-operator?pullRequest=188) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: koikonom<br/>:x: birdayz<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redpanda-data/redpanda-operator?pullRequest=188) it.</sub>

RafalKorepta commented on src/go/k8s/internal/controller/redpanda/cluster_controller.go:
```diff
@@ -1134,3 +1165,68 @@ func isRedpandaClusterVersionManaged(
+		Namespace: observedCluster.Namespace,
+		Name:      observedCluster.Name,
+	}
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
```
I think it would be better to use client.Patch rather than retry mechanism, but in our case it shouldn't matter.

chrisseto commented on src/go/k8s/internal/controller/redpanda/cluster_controller.go:
```diff
@@ -135,6 +135,37 @@ func (r *ClusterReconciler) Reconcile(
+			if cond.Status == corev1.ConditionTrue && err != nil {
+				cond.Status = corev1.ConditionFalse
+				cond.Reason = "ReconcilerError"
+				cond.Message = err.Error()
```
I would actually recommend not updating the status in the case of an unhandled error. The error should be present in the logs and the presence of an error would indicate that the generation _was not_ properly observed.

chrisseto commented on src/go/k8s/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -471,13 +474,15 @@ type ClusterCondition struct {
 const (
 	// ClusterConfiguredConditionType indicates whether the Redpanda cluster configuration is in sync with the desired one
 	ClusterConfiguredConditionType ClusterConditionType = "ClusterConfigured"
+	// ClusterQuiescentConditionType indicates that the controller has no outstanding work to do, based on the observedGeneration.
```
I know this doesn't follow the naming convention but I think we'll be much better served if this is _not_ prefixed with `Cluster`. If `ClusterConfigured` means the "Redpanda Cluster has been configured", then one would assume that `ClusterQuiescent` means the "Redpanda Cluster is quiescent" which is not the case here. This condition pertains to the controller/reconciler itself. Perhaps `ControllerQuiescent` or `ClusterControllerQuiescent`?

sbocinec commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -1210,36 +1216,20 @@ func (s *ClusterStatus) SetRestarting(restarting bool) {
-// It returns 1 when not initialized (as fresh clusters start from 1 replica)
-func (r *Cluster) GetCurrentReplicas() int32 {
-	if r == nil {
+func (r *Cluster) GetReplicas() int32 {
```
nit: comment for this public method is missing

birdayz replied:
done

sbocinec commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -1418,3 +1408,155 @@ func (r *Cluster) GetDecommissionBrokerID() *int32 {
 }
+
+// getNodePoolsFromSpec returns the NodePools defined in the spec.
+// This contains the primary NodePool (driven by spec.replicas for example), and the
```
nit: does "primary" refers to the "default" / "legacy" node pool here? If yes, maybe we should use "default" or "legacy" instead

birdayz replied:
fixed

sbocinec commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -1418,3 +1408,155 @@ func (r *Cluster) GetDecommissionBrokerID() *int32 {
+
+		for i := range sts.Spec.VolumeClaimTemplates {
+			vct := sts.Spec.VolumeClaimTemplates[i]
+			if vct.Name == "datadir" {
+				datadirVcCapacity = vct.Spec.Resources.Requests[corev1.ResourceStorage]
+				if vct.Spec.StorageClassName != nil {
+					datadirVcStorageClassName = *vct.Spec.StorageClassName
+				}
+			}
+			if vct.Name == "shadow-index-cache" {
```
question: Does it makes sense to expose the const names from the statefulset pkg and use it here to avoid failure if these are renamed in the STS https://github.com/redpanda-data/redpanda-operator/blob/main/operator/pkg/resources/statefulset.go#L64-L65 ?

birdayz replied:
good idea, done

sbocinec commented on operator/api/vectorized/v1alpha1/cluster_webhook.go:
```diff
@@ -770,6 +767,32 @@ func (r *Cluster) validateRedpandaMemory() field.ErrorList {
+		requests := np.Resources.Requests.DeepCopy()
+		requests.Cpu().RoundUp(0)
+		requestedCores := requests.Cpu().Value()
+		if np.Resources.Requests.Memory().Value() < requestedCores*MinimumMemoryPerCore {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec").Child("nodePool").Child("resources").Child("requests").Child("memory"),
+					np.Resources.Requests.Memory(),
+					"requests.memory < 2Gi per core; either decrease requests.cpu or increase requests.memory"))
+		}
+
+		redpandaCores := np.Resources.RedpandaCPU().Value()
+		minimumMemoryPerCore := int64(math.Floor(MinimumMemoryPerCore * RedpandaMemoryAllocationRatio))
+		if !np.Resources.RedpandaMemory().IsZero() && np.Resources.RedpandaMemory().Value() < redpandaCores*minimumMemoryPerCore {
+			allErrs = append(allErrs,
+				field.Invalid(
+					field.NewPath("spec").Child("resources").Child("redpanda").Child("memory"),
+					r.Spec.Resources.Requests.Memory(),
+					"redpanda.memory < 2Gi per core; either decrease redpanda.cpu or increase redpanda.memory"))
```
Ah, I see, it's not included neither in the original errors for "defualt" nodepool, so it's good to keep it consistent and possibly improve it later if it turns out to be missing

birdayz replied:
keeping it as-is for now

sbocinec commented on operator/pkg/resources/statefulset_update.go:
```diff
@@ -53,6 +56,21 @@ var (
+	if !ok {
+		npStatus = vectorizedv1alpha1.NodePoolStatus{
+			CurrentReplicas: *r.nodePool.Replicas,
+		}
```
question: why are setting the currentReplicas in the getStatus method? It smells to me on a first look as I don't fully understand the motivation behind.

birdayz replied:
it needs to get initialized at some point. this way, i can control when this happens. every access to status goes through getStatus, and getStatus lazily initializes the nodePoolStatus if necessary.

RafalKorepta commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -1418,3 +1419,155 @@ func (r *Cluster) GetDecommissionBrokerID() *int32 {
+			vct := sts.Spec.VolumeClaimTemplates[i]
+			if vct.Name == "datadir" {
+				datadirVcCapacity = vct.Spec.Resources.Requests[corev1.ResourceStorage]
+				if vct.Spec.StorageClassName != nil {
+					datadirVcStorageClassName = *vct.Spec.StorageClassName
+				}
```
NIT:
```suggestion
				datadirVcStorageClassName = ptr.Deref(vct.Spec.StorageClassName, "")
```

birdayz replied:
done

RafalKorepta commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -1418,3 +1419,155 @@ func (r *Cluster) GetDecommissionBrokerID() *int32 {
+			if vct.Name == "shadow-index-cache" {
+				cacheVcExists = true
+				cacheVcCapacity = vct.Spec.Resources.Requests[corev1.ResourceStorage]
+				if vct.Spec.StorageClassName != nil {
+					cacheVcStorageClassName = *vct.Spec.StorageClassName
+				}
```
NIT:
```suggestion
				cacheVcStorageClassName = ptr.Deref(vct.Spec.StorageClassName, "")
```

birdayz replied:
done

RafalKorepta reviewed (commented):
It would be good to describe how `Spec.Replicas` and `Spec.NodePools[*].Replicas` should be used and calculated.

How `Spec.NodePools[*].Replicas` will behave if only `Spec.Replicas` is used and vice versa.

RafalKorepta commented on operator/pkg/resources/statefulset_scale.go:
```diff
@@ -71,85 +74,135 @@ const (
-	err = r.Status().Update(ctx, r.pandaCluster)
-	if err != nil {
-		return err
-	}
-	if *targetBroker == nonExistantBroker {
+		// Previously, this was accepted, and status.decommissionBrokerID was cleared.
+		// However, this is dangerous.
+		// Broker ID could be missing for a variety of reasons? If we just ignore this, and remove the pod anyway and skip decom'ing..we may remove pod w/ ghost broker left.
+		// Instead, we only decommission, if both pod and brokerID are present. This effectively means, our "state machine" requires upscaling to have finished (pod started + broker registered), before we consider downscaling.
```
Could you correct the whole comment as the first 2 lines are left from previous logic.

It would be good to elaborate about dangerous scenarios.

```
if both pod and brokerID are present.
```

I would change to indicate that BrokerID is an annotation within Pod metadata.

RafalKorepta commented on operator/pkg/resources/testdata/Complex_struct_in_additional_configuration.golden:
```diff
@@ -5,6 +5,9 @@ redpanda:
         - host:
             address: cluster-0.cluster.local
             port: 33145
+        - host:
+            address: cluster-first-0.cluster.local
+            port: 33145
```
Why seed server list have more hosts if only single Node Pool is set? Shouldn't the old host (short name of the pod) be removed?

The `first` node pool is consider second node pool as the `default` is is not included in the `Spec.NodePools`, but it is implied from `Spec.Replicas`, right?

If it's not then what would be the name of the pods? I'm worry that startup will be longer due to first host DNS name being not available in the cluster.

birdayz replied:
this is correct. Spec.Replicas drives the default nodepool "cluster" (with no suffix), and the nodepool provided is an additional STS.

RafalKorepta commented on operator/tests/e2e-with-flags/nodepools/01-two-nodepools.yaml:
```diff
@@ -0,0 +1,45 @@
+      dump-memory-diagnostics-on-alloc-failure-kind: all
+      abort-on-seastar-bad-alloc: ''
+  resources: {}
+status:
+  nodePools:
+    nodepool1:
+      replicas: 0
+    nodepool1:
+      replicas: 3
```
Status should be removed from the test case definition. It can be only asserted (in my opinion). Kuttle, as far as I know` will not update status.

birdayz replied:
done

RafalKorepta commented on operator/tests/e2e-with-flags/nodepools/02-scale-down-first.yaml:
```diff
@@ -0,0 +1,38 @@
+  image: "localhost/redpanda"
+  version: "dev"
+  nodePools:
+    - name: nodepool1
```
Why there are 2 files `02-remove-first.yaml` and `02-scale-down-first.yaml`? I think the one should be removed.

birdayz replied:
good catch. fixed

RafalKorepta commented on operator/tests/e2e-with-flags/nodepools/05-assert.yaml:
```diff
@@ -0,0 +1,4 @@
+apiVersion: kuttl.dev/v1beta1
+kind: TestAssert
+collectors:
+  - command: ../../../hack/get-redpanda-info.sh
```
NIT: The number `05` should be `03` in the fole name. That would indicate the step within test case.

birdayz replied:
fixed

RafalKorepta commented on operator/tests/e2e-with-flags/nodepools/05-clean.yaml:
```diff
@@ -0,0 +1,31 @@
+apiVersion: kuttl.dev/v1beta1
```
NIT: The number `05` should be `03` in the fole name. That would indicate the step within test case.

birdayz commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -1210,36 +1216,20 @@ func (s *ClusterStatus) SetRestarting(restarting bool) {
-func (r *Cluster) GetCurrentReplicas() int32 {
-	if r == nil {
+func (r *Cluster) GetReplicas() int32 {
+	nps := r.getNodePoolsFromSpec() // FIXME - may want to use nodePools-with-removed
```
fixed

birdayz commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -1210,36 +1216,31 @@ func (s *ClusterStatus) SetRestarting(restarting bool) {
+
+	var npNamesSeen []string
+
+	for i := range nps {
+		np := nps[i]
```
hm, the linter complained. is the operator on 1.22 yet? i don't mind to change if, it the linter stops flagging it.

birdayz commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -1418,3 +1419,155 @@ func (r *Cluster) GetDecommissionBrokerID() *int32 {
+	return result
+}
+
+func (r *Cluster) GetNodePools(ctx context.Context, k8sClient client.Reader) ([]*NodePoolSpecWithDeleted, error) {
```
fixed, ty

birdayz commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -1418,3 +1419,155 @@ func (r *Cluster) GetDecommissionBrokerID() *int32 {
+	// Also add "virtual NodePools" based on StatefulSets found.
+	// These represent deleted NodePools - they will not show up in spec.NodePools.
+	var stsList appsv1.StatefulSetList
+	err := k8sClient.List(context.TODO(), &stsList)
```
i think this should work. good call

birdayz replied:
it's actually quite hard..import cycle :/

birdayz replied:
fixed

birdayz replied:
fixed. i changed package structure slightly

birdayz commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -1418,3 +1419,155 @@ func (r *Cluster) GetDecommissionBrokerID() *int32 {
+	var nodePoolsWithDeleted []*NodePoolSpecWithDeleted
+
+	nps := r.getNodePoolsFromSpec()
+	for i := range nps {
+		np := nps[i]
```
linter didn't like it

birdayz commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -1418,3 +1419,155 @@ func (r *Cluster) GetDecommissionBrokerID() *int32 {
+		if strings.EqualFold(r.Name, sts.Name) {
+			npName = nodepools.DefaultNodePoolName
+		} else {
+			npName = sts.Name[len(r.Name)+1:]
```
i changed it a little, also using now TrimPrefix. what do you think?

RafalKorepta replied:
Good

birdayz commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -1418,3 +1419,155 @@ func (r *Cluster) GetDecommissionBrokerID() *int32 {
+		replicas := sts.Spec.Replicas
+		if st, ok := r.Status.NodePools[npName]; ok {
+			replicas = &st.CurrentReplicas
+		}
```
this is a good question. let me check if we can remove it, i am not 100% sure anymore why it is necessary, but i think it is for a reason. i think it's because the normal way of translating currentReplicas to replicas does not work with these "virtual" pools, but i will double check.

birdayz commented on operator/internal/controller/vectorized/cluster_controller.go:
```diff
@@ -510,52 +506,85 @@ func (r *ClusterReconciler) reportStatus(
+		readyReplicas += sts.LastObservedState.Status.ReadyReplicas
+		replicas += sts.LastObservedState.Status.Replicas
+
+		oldNps := redpandaCluster.Status.NodePools[sts.GetNodePool().Name]
```
i need to double check if this is correct. even if there is no status yet for this NP, we do want to create one. the existing code will just do that, as it will have a status struct with default values returned

birdayz commented on operator/internal/controller/vectorized/cluster_controller.go:
```diff
@@ -510,52 +506,85 @@ func (r *ClusterReconciler) reportStatus(
+			return errNonexistentLastObservedState
+		}
+
+		readyReplicas += sts.LastObservedState.Status.ReadyReplicas
+		replicas += sts.LastObservedState.Status.Replicas
```
i do not quite understand. you'd use status.nodepools[*].replicas instead?

birdayz commented on operator/internal/controller/vectorized/cluster_controller.go:
```diff
@@ -902,10 +932,16 @@ func (r *ClusterReconciler) decommissionGhostBrokers(c context.Context, vCluster
+		return p.Status.Phase != corev1.PodRunning
+	})
+
+	if int32(len(pods.Items)) != vCluster.Status.Replicas {
+		log.Info("skipping ghost broker detection, because not all pods are ready.",
+			"replica-number", vCluster.Status.Replicas,
 			"pod-len", len(pods.Items))
+		return
```
It does

birdayz commented on operator/internal/controller/vectorized/cluster_controller_configuration.go:
```diff
@@ -102,14 +102,16 @@ func (r *ClusterReconciler) reconcileConfiguration(
-		log.Info("Centralized configuration hash has changed")
-		if err = statefulSetResource.SetCentralizedConfigurationHashInCluster(ctx, hash); err != nil {
-			return errorWithContext(err, "could not update config hash on statefulset")
+	for _, statefulSetResource := range statefulSetResources {
+		hash, hashChanged, err := r.checkCentralizedConfigurationHashChange(ctx, redpandaCluster, config, schema, lastAppliedConfiguration, statefulSetResource)
```
i added this change, however is it even necessary   to add this? we're not going to put a nil value into this slice.

birdayz commented on operator/internal/controller/vectorized/test/cluster_controller_scale_test.go:
```diff
@@ -212,34 +187,51 @@ var _ = Describe("Redpanda cluster scale resource", func() {
+			Eventually(resourceDataGetter(npFirstKey, &sts, func() interface{} {
 				return *sts.Spec.Replicas
-			}), timeout, interval).Should(Equal(int32(2)))
-			Eventually(statefulSetReplicasReconciler(ctrl.Log, key, redpandaCluster), timeout, interval).Should(Succeed())
-			Eventually(resourceDataGetter(key, redpandaCluster, func() interface{} {
-				return redpandaCluster.GetDecommissionBrokerID()
-			}), timeout, interval).Should(BeNil())
+			}), timeout, interval).Should(Equal(int32(3)))
+			Eventually(resourceDataGetter(npSecondKey, &sts, func() interface{} {
+				return *sts.Spec.Replicas
+			}), timeout, interval).Should(Equal(int32(3)))
```
tbh i did not write this part of the code, but took it from the person who started this :) the checks are still fine i guess; as it checks for the STS being created (which is a different check than looking for the CRD status change)

RafalKorepta replied:
Ok I might missed the fact that test is comparing different statefuleset each time.

birdayz commented:
@RafalKorepta thanks a lot for the thorough comments! i addresses most, and left a few questions.

birdayz commented:
kind ping for review @chrisseto

chrisseto commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -1214,36 +1216,31 @@ func (s *ClusterStatus) SetRestarting(restarting bool) {
-// It returns 1 when not initialized (as fresh clusters start from 1 replica)
-func (r *Cluster) GetCurrentReplicas() int32 {
-	if r == nil {
+func (r *Cluster) SumNodePoolReplicas() int32 {
```
nit: This doesn't indicate if it's returning the current replicas or desired replicas. Please add a comment or rename the method.

birdayz replied:
renamed to GetDesiredReplicas. it looks like it in the diff, but this function is not a replacement for GetCurrentReplicas - this can be seen in metric_controller.go, where a reference to Spec.Replicas is replaced with a call to this function here.

chrisseto commented on operator/api/vectorized/v1alpha1/cluster_types.go:
```diff
@@ -1422,3 +1419,41 @@ func (r *Cluster) GetDecommissionBrokerID() *int32 {
 }
+
+// getNodePoolsFromSpec returns the NodePools defined in the spec.
+// This contains the primary NodePool (driven by spec.replicas for example), and the
```
```suggestion
// This contains the default NodePool (driven by spec.replicas for example), and the
```

Not sure what the terminology is but let's be consistent. If the primary pool has the name "default", let's rename the constant otherwise update this comment.

birdayz replied:
Right, this is an oversight. default nodepool is the correct terminology. Updated the comment

chrisseto commented on operator/pkg/nodepools/pools.go:
```diff
@@ -0,0 +1,143 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+// Package nodepools exists currently because it is required to be a separate due to import conflicts.
```
Please specify the exact import conflict / cycle so we can easily verify when this can be removed.

birdayz replied:
added comment

chrisseto commented on operator/pkg/nodepools/pools.go:
```diff
@@ -0,0 +1,143 @@
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
```
nit: extra newline (we'll need to reconfigure out linters).

chrisseto commented on operator/pkg/nodepools/pools.go:
```diff
@@ -0,0 +1,143 @@
+	corev1 "k8s.io/api/core/v1"
+)
+
+func GetNodePools(ctx context.Context, cluster *vectorizedv1alpha1.Cluster, k8sClient client.Reader) ([]*vectorizedv1alpha1.NodePoolSpecWithDeleted, error) {
```
Seems like this function is doing a lot more than just getting node pools. Please add a comment indicating what's being computed here.

birdayz replied:
added comment

chrisseto commented on operator/pkg/nodepools/pools.go:
```diff
@@ -0,0 +1,143 @@
+	// Also add "virtual NodePools" based on StatefulSets found.
+	// These represent deleted NodePools - they will not show up in spec.NodePools.
+	var stsList appsv1.StatefulSetList
+	err := k8sClient.List(ctx, &stsList, &client.ListOptions{
```
Seems like this should be performed outside of the `nps` loop?

birdayz replied:
this code is not inside any loop.

chrisseto replied:
My mistake, sorry!

chrisseto commented on operator/pkg/nodepools/pools.go:
```diff
@@ -0,0 +1,143 @@
+		})
+	}
+
+	// Also add "virtual NodePools" based on StatefulSets found.
```
This comment is out of place. Either place it in the function comment or directly above the relevant code.

birdayz replied:
Is it out of place? It is right above the statefulset lookup,and that is for sure part of this creation of the "virtual NodePools".

chrisseto replied:
AH. Wow, I really misread this function the first time around. I think what's really throwing me for a loop (aside from using side by side view in github) is the inconsistency of breaking loops within this loop and the number at that.

Could we pick a single method of iterating  and short circuiting be it labeled breaks, `slices.Contains`, or a helper `findFunc[T any]([]T, func(T) bool) (*T, bool)` function?

chrisseto commented on operator/pkg/nodepools/pools.go:
```diff
@@ -0,0 +1,143 @@
+		return nil, err
+	}
+outer:
+	for i := range stsList.Items {
```
Looks like this would be much more readable if it was refactored to not be `O(S*N)`. Based on the string checks on line 61, you could index the STSs by their name and then loop over node pools and do a key lookup which should eliminate the check on 71.

birdayz replied:
to be honest, i am not convinced by either the ergonomics of this refactoring, nor the complexity. these are slices with < 10 items, nobody is going to have hundreds or thousands of node pools. a lookup on a map is not much faster - if faster at all, if the slice has only very little items. an array index lookup is always faster than a map lookup (direct memory access vs extra compute by map, with involves hashing, separate chaining or whatever the implementation does), it only wins with a sufficiently large number of N. and N is going to be super small all the time.

chrisseto replied:
I think I was misunderstanding how much loop nesting was going on here 😓

chrisseto commented on operator/pkg/resources/configmap.go:
```diff
@@ -807,30 +813,59 @@ func (r *ConfigMapResource) SetLastAppliedConfigurationInCluster(
+	r.logger.Info("New seed list")
+	var addresses []string
+
+	for npName, npStatus := range r.pandaCluster.Status.NodePools {
```
Why is this being pulled from Status? This screams race condition to me. Can we not iterate over the nodepools in the spec?

birdayz replied:
status contains also deleted node pools. therefore, we want to include it as input for computing the seed server list.

it is an imperfect solution, that node pools are stored in spec, and not a separate CRD. we discussed this ~1.5 months back in slack. i believe as the operators converge, v2/v3 can introduce a nodepool CRD and do this better - then these cases will disappear, as a separate CRD can distinguish between "delete started" and "finally deleted".

i suppose, for clarity of the code, this function can be refactored to use "spec first", and add add nodePoolStatus-sourced pools on top - looking into refactoring this now.

birdayz replied:
It looks like this special handling is not required here anymore. GetNodePools includes deleted ones, so no need to use status here. Updated the code (awaiting CI to run the entire test suite to confirm this is correct).

chrisseto replied:
> therefore, we want to include it as input for computing the seed server list.

Why do we need to include the (potentially) deleted nodepools in the seed list? Shouldn't it be sufficient to have a subset of all brokers in the seedlist?

birdayz replied:
tbh, i am not 100% sure. if it is safe to do so, i would change it. so far, all brokers were included, so i am following this existing practice. 

i suspect there are corner cases, eg. if a change adds and deletes a nodepool at the same time, we do not point to the old brokers anymore, and new nodes cannot join.

chrisseto commented on operator/pkg/resources/statefulset_scale.go:
```diff
@@ -71,85 +74,134 @@ const (
-	if *r.pandaCluster.Spec.Replicas > r.pandaCluster.Status.CurrentReplicas {
-		r.logger.Info("Upscaling cluster", "replicas", *r.pandaCluster.Spec.Replicas)
+	if ptr.Deref(r.nodePool.Replicas, 0) == npCurrentReplicas && !r.nodePool.Deleted {
+		log.V(logger.DebugLevel).Info("No scaling changes required for this nodepool", "replicas", *r.nodePool.Replicas, "spec replicas", *r.LastObservedState.Spec.Replicas) // No changes to replicas, we do nothing here
```
include the node pools name here?

birdayz replied:
`log` contains nodePool name already.

chrisseto commented on operator/pkg/resources/statefulset_scale.go:
```diff
@@ -71,85 +74,134 @@ const (
+		return nil, err
+	}
+	// Filter out those where NodeID != DecomNodeID.
+	podList.Items = slices.DeleteFunc(podList.Items, func(pod corev1.Pod) bool {
```
Is there any reason for this to not be a range loop with an early return?

```go
for _, pod := range pods {
    if cond {
        return pod, nil
    }
}

return nil, nil // Pod not found
```

birdayz replied:
See https://github.com/redpanda-data/redpanda-operator/pull/188/commits/9deaad0a31b64af7f2280ebbb171f32e8dc66143

Refactored it as suggested, good call. also changed param to non-pointer, which makes much more sense imho.

chrisseto commented on operator/pkg/resources/statefulset_update.go:
```diff
@@ -181,6 +221,14 @@ func (r *StatefulSetResource) getPodList(ctx context.Context) (*corev1.PodList,
 		return nil, fmt.Errorf("unable to list pods: %w", err)
 	}
 
+	podList.Items = slices.DeleteFunc(podList.Items, func(pod corev1.Pod) bool {
```
nit: You could just update the LabelSelector to exclude the `NodePoolKey` label instead of performing client side filtering.

birdayz replied:
i would prefer this as well, and attempted to do this initially. 

however, because of how the code in pkg/labels is built, it requires quite some plumbing to have a different selector. a change is required in multiple places; i would prefer to not do a larger refactoring here.

since this is flagged as a nit, i suggest to keep this as-is.

chrisseto commented on operator/pkg/resources/statefulset_update.go:
```diff
@@ -585,19 +666,31 @@ func (r *StatefulSetResource) shouldUpdate(
+func (r *StatefulSetResource) updateRestartingStatus(ctx context.Context, restarting bool) error {
+	npStatus := r.getNodePoolStatus()
+
+	if !reflect.DeepEqual(restarting, npStatus.Restarting) {
```
I realize that this is pre-existing but why are we using DeepEqual to compare booleans?

birdayz replied:
i have no idea. Kept things as-is rather than refactoring more than needed

chrisseto commented on operator/kind-for-cloud.yaml:
```diff
@@ -9,3 +9,9 @@ nodes:
   image: kindest/node:v1.29.8@sha256:d46b7aa29567e93b27f7531d258c372e829d7224b25e3fc6ffdefed12476d3aa
 - role: worker
   image: kindest/node:v1.29.8@sha256:d46b7aa29567e93b27f7531d258c372e829d7224b25e3fc6ffdefed12476d3aa
+- role: worker
+  image: kindest/node:v1.29.8@sha256:d46b7aa29567e93b27f7531d258c372e829d7224b25e3fc6ffdefed12476d3aa
+- role: worker
+  image: kindest/node:v1.29.8@sha256:d46b7aa29567e93b27f7531d258c372e829d7224b25e3fc6ffdefed12476d3aa
+- role: worker
+  image: kindest/node:v1.29.8@sha256:d46b7aa29567e93b27f7531d258c372e829d7224b25e3fc6ffdefed12476d3aa
```
+1 to Rafal. This isn't super sustainable. Can we not just schedule multiple brokers onto a single node for the purpose of testing?

birdayz replied:
i do not have a strong opinion about this, as long as the test works with either setup. i just don't know if kubelet really puts a lot of pressure on the test machines. have we seen problems here? 

how would i exactly achieve what you suggest? happy to change it in any direction..

chrisseto replied:
Any additional kind nodes are going to come with additional load which in turn means slower CI times and ours are already pretty lengthy as is. I guess I'm not entirely certain of what you need the extra nodes for? If you need to be able to schedule multiple redpanda instances on a single node all you have to do is remove the antiaffinity.

chrisseto commented on operator/pkg/admin/admin.go:
```diff
@@ -17,11 +17,14 @@ import (
 	"time"
 
+	corev1 "k8s.io/api/core/v1"
+
```
nit: extra newline

birdayz replied:
fixed

chrisseto commented on operator/pkg/labels/labels.go:
```diff
@@ -30,6 +30,11 @@ const (
+	// NodePoolKey is used to document the node pool associated with the StatefulSet.
+	NodePoolKey = "cluster.redpanda.com/nodepool"
+
+	// PodLabelNodeIDKey
```
nit: incomplete comment. Seems like `PodNodeIDKey` would be more consistent with the naming convention here? `Label` is already implied by the package.

birdayz replied:
done

chrisseto commented on operator/pkg/labels/labels.go:
```diff
@@ -30,6 +30,11 @@ const (
 	PartOfKey = "app.kubernetes.io/part-of"
 	// The tool being used to manage the operation of an application
 	ManagedByKey = "app.kubernetes.io/managed-by"
+	// NodePoolKey is used to document the node pool associated with the StatefulSet.
```
```suggestion
	// NodePoolKey is used to denote the node pool associated with the StatefulSet.
```

Struggling to explain why document isn't the right word here. I think document implies a separation of information and thing that's being informed about. A label on something wouldn't document something about itself but rather denote something about itself.

birdayz replied:
updated

chrisseto commented:
Thanks for the ping :)

birdayz commented:
Thanks for the review! Addressed most of the comments.
Ready for the next round.

Mostly, these are open and awaiting further feedback:
- https://github.com/redpanda-data/redpanda-operator/pull/188#discussion_r1833414738 - Help required how to actually achieve this (no specific opinion on my side)
- https://github.com/redpanda-data/redpanda-operator/pull/188#discussion_r1833397989 is probably most controversial, i added an explanation why i believe it can stay as-is

birdayz commented:
@chrisseto kind ping for next round. would be really important to bring this over the finish line this week!

birdayz commented:
Added another commit on top, that allows us to have an offset for the hostIndex per nodepool.

endpoint templates use hostIndex - which is equal to pod ordinal - to build advertised listener addresses. however, with multiple STS, it will clash.
by having an offset, we can provide separate port ranges per NP/STS.

See https://github.com/redpanda-data/redpanda-operator/pull/188/commits/ab001c353b468a033b24f54b1e0aa282c24bba15

chrisseto commented on operator/pkg/nodepools/pools.go:
```diff
@@ -0,0 +1,153 @@
+	for i := range stsList.Items {
+		sts := stsList.Items[i]
+
+		for _, ownerRef := range sts.OwnerReferences {
```
I find this check and the later check for the existence of a `redpanda` container a bit odd. We're already using a label selector to filter our list of Sts's; is there a reason we're performing these extra checks? Are they checking for different things? If they're not, why does one error but the other does not?

I like to avoid "anxious" code when possible as it creates an ethical dilemma  when it comes time to refactor. I'll have all the above questions with potentially no one to answer them nor test cases to validate them.

If these checks are in place for a reason, document what exactly they're checking for and ideally the situations in which they might trigger.

If these checks are just anxiety or paranoia that you can't shake, group them together and label them as such. A simple `// Paranoid check to make sure we don't delete other STSs` can save so much head-aching in the future.

chrisseto reviewed (approved):
I can't seem to find the comment again but I'm still curious about the management of the seed servers list and why it's populated from Status instead of the Spec. 

Before merging could you either point me to the answer or comment again?

birdayz commented:
> I can't seem to find the comment again but I'm still curious about the management of the seed servers list and why it's populated from Status instead of the Spec.

based on your comment, it refactored it. It's now not based on status anymore.
See commit: https://github.com/redpanda-data/redpanda-operator/pull/188/commits/7fce0e8c4788c3afc576835c9d79f820227273ff

It used to use status, because spec does not contain deleted nodepools. however, later, a helper function was added, that always returns all nodepools, including these "virtual pools" based on the deleted ones. now, we're using that, status is not required anymore.
